File: GdPicture/WinForm/C#/Text Writer/Form1.cs
using GdPicture14; 

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
using GdPicture14;

namespace Projet1
{
    internal partial class Form1 : System.Windows.Forms.Form 
    { 
       
        
        private string POS_INDICATOR = System.Convert.ToString( System.Convert.ToChar( 32 ) ) + System.Convert.ToChar( 182 ); 
        private int m_ImageID, _nClonedImageID; 
        private int _nStartPosX, _nStartPosY; 
        private bool _bCursorEnabled; 
        private string _sText; 
        private string _sFontName;
        private System.Drawing.FontStyle _fFontStyle; 
        private GdPictureImaging oGdPictureImaging = new GdPictureImaging(); 
        public   System.Windows.Forms.Label oCursor; 
        
        
        
        
        
        private void btZoomIN_Click( System.Object eventSender, System.EventArgs eventArgs ) 
        {
            if (GdViewer1.ZoomIN() != GdPictureStatus.OK)
                MessageBox.Show("Cannot Zoom In");
        } 
        
        
        
        private void btZoomOUT_Click( System.Object eventSender, System.EventArgs eventArgs ) 
        {
            if (GdViewer1.ZoomOUT() != GdPictureStatus.OK)
                MessageBox.Show("Cannot Zoom OUT"); 
        } 
        
        
        
        private void Form1_Load( System.Object eventSender, System.EventArgs eventArgs ) 
        { 
            this.oCursor = new System.Windows.Forms.Label(); 
            oCursor.BackColor = System.Drawing.SystemColors.ActiveCaption; 
            oCursor.Location = new System.Drawing.Point( 645, 500 ); 
            oCursor.Name = "oCursor"; 
            oCursor.Size = new System.Drawing.Size( 1, 10 ); 
            oCursor.TabIndex = 0; 
            
            GdViewer1.Controls.Add( oCursor ); 
            oCursor.Visible = false; 
            _bCursorEnabled = false;

            LicenseManager oLicenseManager = new LicenseManager(); //Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
            oLicenseManager.RegisterKEY("XXXX"); //Please, replace XXXX by a valid demo or commercial license key.

            GdViewer1.ZoomMode = ViewerZoomMode.ZoomMode100; 
            m_ImageID = oGdPictureImaging.CreateNewGdPictureImage( 800, 500, 32, Color.White );
            if (m_ImageID == 0)
            {
                MessageBox.Show("Error: " + oGdPictureImaging.GetStat().ToString());
                return;
            }
            if (GdViewer1.DisplayFromGdPictureImage(m_ImageID) != GdPictureStatus.OK)
                MessageBox.Show("Cannot Display Message");
            GdViewer1.Cursor = Cursors.IBeam; 
            GdViewer1.KeepDocumentPosition = true; 
            GdViewer1.MouseMode = ViewerMouseMode.MouseModeDefault; 
            _sFontName = "Arial"; 
            
            
        } 
        
        
        
        private void Form1_FormClosed( System.Object eventSender, System.Windows.Forms.FormClosedEventArgs eventArgs ) 
        { 
            oGdPictureImaging.ReleaseGdPictureImage( m_ImageID ); 
        } 
        
        
        
        
        private void BurnText( string sText, bool bBurnPosIndicator ) 
        { 
            if ( _nClonedImageID != 0 )
            { 
                oGdPictureImaging.ReleaseGdPictureImage( _nClonedImageID ); 
            } 
            _nClonedImageID = oGdPictureImaging.CreateClonedGdPictureImage( m_ImageID );
            if (_nClonedImageID == 0)
            {
                MessageBox.Show("Error: " + oGdPictureImaging.GetStat().ToString());
                return;
            }
            
            if ( bBurnPosIndicator )
            { 
                sText = sText + POS_INDICATOR; 
            }
            if (sText == null)
                return;
            if(oGdPictureImaging.DrawText( _nClonedImageID, sText, _nStartPosX, _nStartPosY, 
                int.Parse( txtFontSize.Text ), _fFontStyle, Color.Black, _sFontName, true ) != GdPictureStatus.OK)
                MessageBox.Show("Error: "+ oGdPictureImaging.GetStat().ToString()); 
            
            if(GdViewer1.DisplayFromGdPictureImage( _nClonedImageID )!= GdPictureStatus.OK)
                MessageBox.Show("Cannot Display Message"); 
        } 
        
        
        
        private void MoveCursor() 
        { 
            int nCursorX = 0, nCursorY = 0; 
            float nTextWidth = 0, nTextHeight = 0; 
            string sCurrentLineOfText = null; 
            int nCurrentLineOfTextPos = 0; 
            float nLineHeight = 0; 
            
            nCurrentLineOfTextPos = Strings.InStrRev( _sText, System.Convert.ToString( System.Convert.ToChar( 10 ) ) + System.Convert.ToChar( 13 ), -1, (Microsoft.VisualBasic.CompareMethod)(0) ); 
            if ( nCurrentLineOfTextPos > 0 ) 
            { 
                sCurrentLineOfText = _sText.Substring( nCurrentLineOfTextPos + 1 ); 
            } 
            else 
            { 
                sCurrentLineOfText = _sText; 
            } 
            
            nTextWidth = oGdPictureImaging.GetTextWidth( m_ImageID, sCurrentLineOfText, _sFontName, int.Parse( txtFontSize.Text ), _fFontStyle ); 
            nTextHeight = oGdPictureImaging.GetTextHeight( m_ImageID, _sText + POS_INDICATOR, _sFontName, int.Parse( txtFontSize.Text ), _fFontStyle ); 
            nLineHeight = oGdPictureImaging.GetTextHeight( m_ImageID, sCurrentLineOfText + POS_INDICATOR, _sFontName, int.Parse( txtFontSize.Text ), _fFontStyle ); 
            
            GdViewer1.CoordDocumentToViewer( System.Convert.ToInt32( _nStartPosX + nTextWidth ), _nStartPosY, ref nCursorX, ref nCursorY ); 
            oCursor.Left = nCursorX; 
            oCursor.Top = System.Convert.ToInt32( nCursorY + ( ( nTextHeight - nLineHeight ) * GdViewer1.Zoom ) ); 
            oCursor.Height = System.Convert.ToInt32( nLineHeight * GdViewer1.Zoom ); 
        } 
        
        
        
        private void SaveState() 
        { 
            BurnText( _sText, false ); 
            oCursor.Visible = false; 
            _bCursorEnabled = false; 
            _sText = ""; 
            if ( _nClonedImageID != 0 ) 
            { 
                oGdPictureImaging.ReleaseGdPictureImage( m_ImageID ); 
                m_ImageID = _nClonedImageID; 
                _nClonedImageID = 0; 
                if(GdViewer1.DisplayFromGdPictureImage( m_ImageID )!= GdPictureStatus.OK)
                MessageBox.Show("Cannot Display Message"); 
            } 
        } 
        
        
        private void gdviewer1_Leave( object sender, System.EventArgs e ) 
        { 
            SaveState(); 
        } 
        
        
        private void Timer1_Tick( System.Object eventSender, System.EventArgs eventArgs ) 
        { 
            if ( _bCursorEnabled ) 
            { 
                oCursor.Visible = !( oCursor.Visible ); 
            } 
        } 
        
        
        private void GdViewer1_KeyPress( object sender, System.Windows.Forms.KeyPressEventArgs e ) 
        { 
            if ( _bCursorEnabled ) 
            { 
                
                if ( System.Convert.ToInt32( e.KeyChar ) != 9 ) 
                { // Exclude TAB
                
                    switch ( System.Convert.ToInt32( e.KeyChar ) ) 
                    {
                        case 8:
                            if ( _sText.Length > 0 ) 
                            { 
                                if ( char.Parse( _sText.Substring( _sText.Length-1 ) ) == System.Convert.ToChar( 13 ) ) 
                                { 
                                    if ( _sText.Length - 3 > 0 ) 
                                    { 
                                        _sText = Strings.Mid( _sText, 1, _sText.Length - 3 ); 
                                    } 
                                    else 
                                    { 
                                        _sText = ""; 
                                    } 
                                } 
                                else 
                                { 
                                    _sText = Strings.Mid( _sText, 1, _sText.Length - 1 ); 
                                } 
                            } 
                            break;
                        case 13:
                            _sText = _sText + System.Convert.ToChar( 10 ) + System.Convert.ToChar( 13 ); 
                            break;
                        default:
                            _sText = _sText + e.KeyChar; 
                            break;
                    }
                    
                    BurnText( _sText, true ); 
                } 
            } 
        } 
        
        
        
        private void GdViewer1_MouseDown( object sender, System.Windows.Forms.MouseEventArgs e ) 
        { 
            SaveState();

            _fFontStyle = FontStyle.Regular; 
            if ( System.Convert.ToDouble( chkBold.CheckState ) == 1 )
            {
                _fFontStyle = _fFontStyle | FontStyle.Bold; 
            } 
            if ( System.Convert.ToDouble( chkItalic.CheckState ) == 1 )
            {
                _fFontStyle = _fFontStyle | FontStyle.Italic; 
            } 
            if ( System.Convert.ToDouble( chkUnderlined.CheckState ) == 1 )
            {
                _fFontStyle = _fFontStyle | FontStyle.Underline; 
            } 
            
            GdViewer1.CoordViewerToDocument( e.X, e.Y, ref _nStartPosX, ref _nStartPosY ); 
            
            if ( _nStartPosX >= 0 & _nStartPosY >= 0 & _nStartPosX < oGdPictureImaging.GetWidth( m_ImageID ) & _nStartPosY < oGdPictureImaging.GetHeight( m_ImageID ) ) 
            { 
                _bCursorEnabled = true; 
                BurnText( "", true ); 
                MoveCursor(); 
            } 
        } 
        
        
        private void GdViewer1_PageDisplayed() 
        { 
            MoveCursor(); 
        } 
        
        private static Form1 transDefaultFormForm1 = null;
        public static Form1 TransDefaultFormForm1
        { 
        	get
        	{ 
        		if (transDefaultFormForm1 == null)
        		{
        			transDefaultFormForm1 = new Form1();
        		}
        		return transDefaultFormForm1;
        	} 
        }



    } 
    
} 

File: GdPicture/WinForm/vb.net/Text Writer/Form1.vb
Option Strict On
Option Explicit On

Imports GdPicture14

Public Class Form1
    Private POS_INDICATOR As String = Chr(32) & Chr(182)
    Private ImageID, nClonedImageID As Integer
    Private nStartPosX, nStartPosY As Integer
    Private bCursorEnabled As Boolean
    Private sText As String
    Private sFontName As String
    Private FontStyle As FontStyle
    Private nFontHeight As Integer
    Private oGdPictureImaging As New GdPictureImaging
    Public WithEvents oCursor As System.Windows.Forms.Label


    Private Sub Form1_Load(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles MyBase.Load
        Dim oLicenseManager As New GdPicture14.LicenseManager() 'Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
        oLicenseManager.RegisterKEY("XXXX") 'Please, replace XXXX by a valid demo or commercial license key.

        Me.oCursor = New System.Windows.Forms.Label
        oCursor.BackColor = System.Drawing.SystemColors.ActiveCaption
        oCursor.Location = New System.Drawing.Point(645, 500)
        oCursor.Name = "oCursor"
        oCursor.Size = New System.Drawing.Size(1, 10)
        oCursor.TabIndex = 0

        GdViewer1.Controls.Add(oCursor)
        oCursor.Visible = False
        bCursorEnabled = False


        GdViewer1.ZoomMode = ViewerZoomMode.ZoomMode100
        ImageID = oGdPictureImaging.CreateNewGdPictureImage(800, 500, 32, Color.White)
        GdViewer1.DisplayFromGdPictureImage(ImageID)
        GdViewer1.Cursor = Cursors.IBeam
        GdViewer1.KeepDocumentPosition = True
        GdViewer1.MouseMode = ViewerMouseMode.MouseModeDefault
        sFontName = "Arial"
    End Sub


    Private Sub btZoomIN_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles btZoomIN.Click
        GdViewer1.ZoomIN()
    End Sub


    Private Sub btZoomOUT_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles btZoomOUT.Click
        GdViewer1.ZoomOUT()
    End Sub


    Private Sub Form1_FormClosed(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.FormClosedEventArgs) Handles Me.FormClosed
        oGdPictureImaging.ReleaseGdPictureImage(ImageID)
    End Sub


    Private Sub BurnText(ByVal sText As String, ByVal bBurnPosIndicator As Boolean)
        If nClonedImageID <> 0 Then oGdPictureImaging.ReleaseGdPictureImage(nClonedImageID)
        nClonedImageID = oGdPictureImaging.CreateClonedGdPictureImage(ImageID)

        If bBurnPosIndicator Then sText = sText & POS_INDICATOR
        Call oGdPictureImaging.DrawText(nClonedImageID, sText, nStartPosX, nStartPosY, CInt(txtFontSize.Text), FontStyle, Color.Black, sFontName, True)

        GdViewer1.DisplayFromGdPictureImage(nClonedImageID)
    End Sub


    Private Sub MoveCursor()
        Dim nCursorX, nCursorY As Integer
        Dim nTextWidth, nTextHeight As Single
        Dim sCurrentLineOfText As String
        Dim nCurrentLineOfTextPos As Integer
        Dim nLineHeight As Single

        nCurrentLineOfTextPos = InStrRev(sText, Chr(10) & Chr(13))
        If nCurrentLineOfTextPos > 0 Then
            sCurrentLineOfText = Mid(sText, nCurrentLineOfTextPos + 2)
        Else
            sCurrentLineOfText = sText
        End If

        nTextWidth = oGdPictureImaging.GetTextWidth(ImageID, sCurrentLineOfText, sFontName, CInt(txtFontSize.Text), FontStyle)
        nTextHeight = oGdPictureImaging.GetTextHeight(ImageID, sText & POS_INDICATOR, sFontName, CInt(txtFontSize.Text), FontStyle)
        nLineHeight = oGdPictureImaging.GetTextHeight(ImageID, sCurrentLineOfText & POS_INDICATOR, sFontName, CInt(txtFontSize.Text), FontStyle)

        Call GdViewer1.CoordDocumentToViewer(CInt(nStartPosX + nTextWidth), nStartPosY, nCursorX, nCursorY)
        oCursor.Left = nCursorX
        oCursor.Top = CInt(nCursorY + ((nTextHeight - nLineHeight) * GdViewer1.Zoom))
        oCursor.Height = CInt(nLineHeight * GdViewer1.Zoom)
    End Sub


    Private Sub SaveState()
        Call BurnText(sText, False)
        oCursor.Visible = False
        bCursorEnabled = False
        sText = ""
        If nClonedImageID <> 0 Then
            oGdPictureImaging.ReleaseGdPictureImage(ImageID)
            ImageID = nClonedImageID
            nClonedImageID = 0
            GdViewer1.DisplayFromGdPictureImage(ImageID)
        End If
    End Sub


    Private Sub gdviewer1_Leave(ByVal sender As Object, ByVal e As System.EventArgs) Handles GdViewer1.Leave
        SaveState()
    End Sub


    Private Sub Timer1_Tick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Timer1.Tick
        If bCursorEnabled Then
            oCursor.Visible = Not oCursor.Visible
        End If
    End Sub


    Private Sub GdViewer1_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles GdViewer1.KeyPress
        If bCursorEnabled Then

            If Asc(e.KeyChar) <> 9 Then 'Exclude TAB
                Select Case Asc(e.KeyChar)
                    Case 8
                        If Len(sText) > 0 Then
                            If Mid(sText, Len(sText)) = Chr(13) Then
                                If Len(sText) - 3 > 0 Then
                                    sText = Mid(sText, 1, Len(sText) - 3)
                                Else
                                    sText = ""
                                End If
                            Else
                                sText = Mid(sText, 1, Len(sText) - 1)
                            End If
                        End If
                    Case 13
                        sText = sText & Chr(10) & Chr(13)
                    Case Else
                        sText = sText & e.KeyChar
                End Select
                Call BurnText(sText, True)
            End If
        End If
    End Sub


    Private Sub GdViewer1_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles GdViewer1.MouseDown
        SaveState()

        FontStyle = FontStyle.Regular
        If chkBold.CheckState = 1 Then FontStyle = FontStyle Or FontStyle.Bold
        If chkItalic.CheckState = 1 Then FontStyle = FontStyle Or FontStyle.Italic
        If chkUnderlined.CheckState = 1 Then FontStyle = FontStyle Or FontStyle.Underline

        Call GdViewer1.CoordViewerToDocument(e.X, e.Y, nStartPosX, nStartPosY)

        If nStartPosX >= 0 And nStartPosY >= 0 And nStartPosX < oGdPictureImaging.GetWidth(ImageID) And nStartPosY < oGdPictureImaging.GetHeight(ImageID) Then
            bCursorEnabled = True
            Call BurnText("", True)
            Call MoveCursor()
        End If
    End Sub


    Private Sub GdViewer1_PageDisplayed() Handles GdViewer1.PageDisplayed
        MoveCursor()
    End Sub
End Class